sql/src/main/java/io/crate/jobs/JobExecutionContext.java

```diff
@@ -99,24 +96,31 @@ private JobExecutionContext(UUID jobId,
                                 String coordinatorNodeId,
                                 Collection<String> participatingNodes,
                                 StatsTables statsTables,
-                                LinkedHashMap<Integer, ExecutionSubContext> contextMap) throws Exception {
+                                List<ExecutionSubContext> orderedContexts) throws Exception {
         this.coordinatorNodeId = coordinatorNodeId;
         this.participatedNodes = participatingNodes;
-        orderedContextIds = Lists.newArrayList(contextMap.keySet());
+        orderedContextIds = new IntArrayList(orderedContexts.size());
         this.jobId = jobId;
         this.statsTables = statsTables;
-        prepare(contextMap);
 
-        subContexts = new ConcurrentHashMap<>(contextMap.size());
-        numSubContexts = new AtomicInteger(contextMap.size());
+        subContexts = new ConcurrentHashMap<>(orderedContexts.size());
+        numSubContexts = new AtomicInteger(orderedContexts.size());
 
-        for (Map.Entry<Integer, ExecutionSubContext> entry : contextMap.entrySet()) {
-            int subContextId = entry.getKey();
-            Futures.addCallback(entry.getValue().completionFuture(), new RemoveSubContextListener(subContextId));
-            subContexts.put(entry.getKey(), entry.getValue());
-            LOGGER.trace("adding subContext {}, now there are {} subContexts", subContextId, subContexts.size());
+        boolean traceEnabled = LOGGER.isTraceEnabled();
+        for (ExecutionSubContext context : orderedContexts) {
+            int subContextId = context.id();
+            orderedContextIds.add(subContextId);
+            Futures.addCallback(context.completionFuture(), new RemoveSubContextListener(subContextId));
+            ExecutionSubContext existingContext = subContexts.put(subContextId, context);
+            if (existingContext != null) {
+                throw new IllegalArgumentException(String.format(Locale.ENGLISH,
```

Comment:
Yes it would be, I had copied it and didn't bother changing it because this is only raised if there is a bug in the planner or contextPreparer.